kaggle_environments/envs/lux_ai_2022/luxai2022/factory.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import annotations
+from collections import deque
+import time
+from typing import List
+import numpy as np
+from luxai2022.config import EnvConfig
+from luxai2022.map.board import Board
+from luxai2022.map.position import Position
+from luxai2022.team import Team
+from luxai2022.unit import UnitCargo
+from luxai2022.actions import move_deltas
+from luxai2022.globals import TERM_COLORS
+try:
+    from termcolor import colored
+except:
+    pass
+
+
+def compute_water_info(init: np.ndarray, MIN_LICHEN_TO_SPREAD: int, lichen: np.ndarray, lichen_strains: np.ndarray, strain_id: int, forbidden: np.ndarray):
+    # TODO - improve the performance here with cached solution
+    frontier = deque(init)
+    seen = set(map(tuple, init))
+    grow_lichen_positions = set()
+    H, W = lichen.shape
+    while len(frontier) > 0:
```

Comment:
`while` loops can cause... issues. I usually write an escape if running in production environment.

```
i = 0
while <x>
i += 1
if i > 10000
    print("oh noes")
    break
```